File: Prototype/Assets/Scripts/MapGenerator/MapGenerator.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Events;
using csDelaunay;

public class MapGenerator : MonoBehaviour
{
    public DiagramLayout layout;
    //public MapFence fence;
    public CellMeshConfig cellMeshConfig;
    public CellFenceConfig cellFenceConfig;

    public GameObject bridgePrefab;
    public float minEdgeLengthForBridge = 5.0f;

    // This is where we will store the resulting data
    private Dictionary<Vector2f, Site> sites;
    private List<Edge> edges;
    private Rectf bounds;


    public UnityEvent OnMapGenerated;

    private void Start()
    {
        //fence.Initialize();

        bounds = layout.GetBounds();
        Voronoi voronoi = new Voronoi(layout.CreatePoints(), bounds);
        voronoi.LloydRelaxation(layout.LloydRelaxIteraions);

        sites = voronoi.SitesIndexedByLocation;
        edges = voronoi.Edges;

        GameObject cellContainer = new GameObject("Cells");
        cellContainer.transform.parent = transform;

        // prepare cells
        List<MapCell> cells = new List<MapCell>();

        foreach (var s in sites.Values)
            if (!IsBorderCell(s))
                cells.Add(CreateCell(s, cellContainer));

        // create bridges and connect references
        GameObject bridgeContainer = new GameObject("Bridges");
        bridgeContainer.transform.parent = transform;
        CreateBridges(cells, bridgeContainer);

        // now generate real cells
        foreach (var s in sites.Values)
        {
            if (!IsBorderCell(s))
            {
                MapCell cell = cells.Find(c => c.CellSiteIndex == s.SiteIndex);
                ConvexPolygon cellPolygon = new ConvexPolygon(s.Region(bounds), s.Coord);
                CellGenerator generator = cell.gameObject.AddComponent<CellGenerator>();
                generator.Generate(
                    cellPolygon,
                    cellMeshConfig,
                    cellFenceConfig
                );
            }
        }

        /*
        CreateTerrain();
        CreateEdges();
        CreateVertices();
        */


        StartCoroutine(coTriggerOnMapGenerated());
    }

    IEnumerator coTriggerOnMapGenerated()
    {
        yield return new WaitForEndOfFrame();
        OnMapGenerated.Invoke();
    }

    MapCell CreateCell(Site cell, GameObject parent)
    {
        GameObject cellObject = new GameObject(cell.SiteIndex.ToString());
        cellObject.transform.parent = parent.transform;
        cellObject.transform.position = new Vector3(cell.Coord.x, 0, cell.Coord.y);

        MapCell mCell = cellObject.AddComponent<MapCell>();
        mCell.CellSiteIndex = cell.SiteIndex;
        return mCell;
    }

    void CreateBridges(List<MapCell> cells, GameObject parent)
    {
        System.Func<Site, MapCell> GetCell = (s) =>
        {
            foreach (var cell in cells)
                if (cell.CellSiteIndex == s.SiteIndex)
                    return cell;
            return null;
        };

        foreach (var edge in edges)
        {
            if (!edge.Visible())
                continue;

            Vector2f leftVertex = edge.ClippedEnds[LR.LEFT];
            Vector2f rightVertex = edge.ClippedEnds[LR.RIGHT];

            if (leftVertex.DistanceSquare(rightVertex) < minEdgeLengthForBridge * minEdgeLengthForBridge)
                continue;

            var cellA = GetCell(edge.LeftSite);
            var cellB = GetCell(edge.RightSite);
            if (cellA == null || cellB == null)
                continue;

            Vector3 position = new Vector3(
                    (leftVertex.x + rightVertex.x) * 0.5f,
                    0.0f,
                    (leftVertex.y + rightVertex.y) * 0.5f
                );

            GameObject bridgeObject = Instantiate(bridgePrefab, parent.transform);
            bridgeObject.transform.position = position;
            bridgeObject.transform.right = (new Vector3(rightVertex.x, 0.0f, rightVertex.y) - position).normalized;
            CellBridge bridge = bridgeObject.GetComponent<CellBridge>();

            bridge.CellA = cellA;
            bridge.CellB = cellB;

            cellA.Bridges.Add(new MapCell.BridgeTo() { Bridge = bridge, Cell = cellB });
            cellB.Bridges.Add(new MapCell.BridgeTo() { Bridge = bridge, Cell = cellA });
        }
    }

    /*
    void CreateCell(Site cell, GameObject parent)
    {
        ConvexPolygon cellPolygon = new ConvexPolygon(cell.Region(layout.GetBounds()), cell.Coord);
        cellPolygon = cellPolygon.ScaledBy(cellMeshConfig.mainScale);

        GameObject cellObject = new GameObject(cell.SiteIndex.ToString());
        cellObject.transform.parent = parent.transform;
        cellObject.transform.position = new Vector3(cell.Coord.x, 0, cell.Coord.y);

        MeshRenderer renderer = cellObject.AddComponent<MeshRenderer>();
        renderer.material = new Material(Shader.Find("Lightweight Render Pipeline/Lit"));

        MeshFilter filter = cellObject.AddComponent<MeshFilter>();
        CellMeshGenerator generator = new CellMeshGenerator(cellMeshConfig, cellPolygon);
        generator.PrepareOutlines();
        generator.PrepareBaseMeshStructure();
        filter.mesh = generator.BuildMesh();
    }
    */

    bool IsBorderCell(Site cell)
    {
        foreach (var p in cell.Region(layout.GetBounds()))
            if (p.x == 0 || p.x == layout.width || p.y == 0 || p.y == layout.height)
                return true;
        return false;
    }

    /*
    void CreateTerrain()
    {
        GameObject terrainObject = new GameObject("Terrain");
        terrainObject.transform.parent = transform;

        TerrainData data = new TerrainData();
        data.size = new Vector3(layout.width, 1, layout.height);
        data.baseMapResolution = 64;
        data.terrainLayers = new TerrainLayer[] {
            new TerrainLayer()
        };
        data.terrainLayers[0].tileSize = new Vector2(20, 20);
        float[,,] alphamap = new float[data.alphamapResolution, data.alphamapResolution,1];
        for (int x = 0; x < data.alphamapResolution; x++)
            for (int y = 0; y < data.alphamapResolution; y++)
                alphamap[x, y, 0] = 1.0f;
        data.SetAlphamaps(0, 0, alphamap);

        Terrain terrain = terrainObject.AddComponent<Terrain>();
        terrain.terrainData = data;
        terrain.materialTemplate = new Material(Shader.Find("Lightweight Render Pipeline/Terrain/Lit"));

        terrainObject.AddComponent<TerrainCollider>().terrainData = data;
    }
    void CreateEdges()
    {
        GameObject edgesContainer = new GameObject("Edges");
        edgesContainer.transform.parent = transform;


        foreach (Edge edge in edges)
        {
            if (!edge.Visible())
                continue;

            GameObject edgeGroup = new GameObject(edge.EdgeIndex.ToString());
            edgeGroup.transform.parent = edgesContainer.transform;

            Vector2f leftVertex = edge.ClippedEnds[LR.LEFT];
            Vector2f rightVertex = edge.ClippedEnds[LR.RIGHT];

            fence.CreateEdgeFence(
                    new Vector3(leftVertex.x, 0, leftVertex.y),
                    new Vector3(rightVertex.x, 0, rightVertex.y),
                    edgeGroup
                );
        }
    }

    void CreateVertices()
    {
        GameObject verticesContainer = new GameObject("Vertices");
        verticesContainer.transform.parent = transform;

        List<Vector2f> spawnedVertices = new List<Vector2f>();

        foreach (Edge edge in edges)
        {
            if (!edge.Visible())
                continue;

            Vector2f leftVertex = edge.ClippedEnds[LR.LEFT];
            Vector2f rightVertex = edge.ClippedEnds[LR.RIGHT];

            if (spawnedVertices.Find(v => v.DistanceSquare(leftVertex) < 0.01f).magnitude == 0)
            {
                fence.CreateVertex(new Vector3(leftVertex.x, 0, leftVertex.y), verticesContainer);
                spawnedVertices.Add(leftVertex);
            }

            if (spawnedVertices.Find(v => v.DistanceSquare(rightVertex) < 0.01f).magnitude == 0)
            {
                fence.CreateVertex(new Vector3(rightVertex.x, 0, rightVertex.y), verticesContainer);
                spawnedVertices.Add(rightVertex);
            }

        }
    }
    */
}

File: Main/Game/Game/MapGenerator/MapGenerator.cpp
#include <random>
#include <memory>
#include <set>
#include <cppDelaunay/delaunay/Edge.h>
#include <glm/glm.hpp>
#include "MapGenerator.h"
#include "CellGenerator.h"
#include "CellSetuper.h"
#include "ext/LloydRelaxation.h"
#include "HFEngine.h"
#include "ECS/Components/Transform.h"

namespace {
	std::vector<Delaunay::Point*> GeneratePoints(DiagramLayout layout)
	{
        static std::random_device rd;
        static std::mt19937 gen(rd());
        static std::uniform_real_distribution<> dis(0.0,1.0);

		std::vector<Delaunay::Point*> points;
        float wPart = layout.width / (float)layout.columns;
        float hPart = layout.height / (float)layout.rows;

        for (int x = 0; x < layout.columns; x++)
        {
            for (int y = 0; y < layout.rows; y++)
            {
                points.push_back(
                    Delaunay::Point::create(
                        glm::mix(wPart * (float)x, wPart * (float)(x + 1), (float)dis(gen)),
                        glm::mix(hPart * (float)y, hPart * (float)(y + 1), (float)dis(gen))
                        )
                    );
            }
        }
        return points;
	}
	Delaunay::Rectangle GetBounds(DiagramLayout layout)
	{
		return Delaunay::Rectangle(0, 0, layout.width, layout.height);
	}
    ConvexPolygon CreateLocalPolygon(Delaunay::Site* cell, Delaunay::Rectangle bounds)
    {
        auto center = cell->coord();
        std::vector<glm::vec2> points;
        for (auto p : cell->region(bounds))
            points.push_back({ p->x - center->x, p->y - center->y });
        return ConvexPolygon(points);
    }
}

void MapGenerator::Generate()
{
    assert(!_generated && "Map already generated");

    bounds = GetBounds(config.layout);
    std::unique_ptr<Delaunay::Voronoi> voronoi;
    std::set<Delaunay::Site*> sites;
    std::vector<GameObject> cells;
    GameObject MapObject;
    GameObject cellContainer;
    GameObject bridgeContainer;

    while (true)
    {
        auto points = GeneratePoints(config.layout);

        voronoi = DelaunayExt::VoronoiWithLloydRelaxation(points, bounds, config.layout.LloydRelaxIteraions);
        edges = voronoi->edges();

        for (auto edge : edges)
        {
            sites.insert(edge->rightSite());
            sites.insert(edge->leftSite());
        }

        MapObject = HFEngine::ECS.CreateGameObject("Map");

        // prepare cells
        cellContainer = HFEngine::ECS.CreateGameObject(MapObject, "Cells");
        for (auto s : sites)
            if (!IsBorderCell(s))
                cells.push_back(CreateCell(s, cellContainer));

        // create bridges and connect references
        bridgeContainer = HFEngine::ECS.CreateGameObject(MapObject, "Bridges");
        CreateBridges(cells, bridgeContainer);

        if (ValidateMapLayout(cells))
        {
            break; // all is fine, continue...
        }
        else
        {
            // somethins is wrong with map layout
            // clear it and start process again
            sites.clear();
            cells.clear();
            HFEngine::ECS.DestroyGameObject(MapObject); // map root object, it will recusively destroy everything below
            voronoi.reset();
        }
    }

    // now generate real cells
    // MOVED TO MapSetuper::Setup


    // and setup those cells
    mapSetuper = std::make_unique<MapSetuper>(config, cells);
    mapSetuper->_debugLiteMode = _debugLiteMode;
    mapSetuper->Setup();

    _generated = true;
}

GameObject MapGenerator::GetStartupCell()
{
    assert(_generated && "Map not generated yet");
    return mapSetuper->GetStartupCell();
}



bool MapGenerator::IsBorderCell(Delaunay::Site* cell)
{
    for (auto p : cell->region(bounds))
        if (p->x == 0 || p->x == config.layout.width || p->y == 0 || p->y == config.layout.height)
            return true;
    return false;
}

GameObject MapGenerator::CreateCell(Delaunay::Site* cell, GameObject parent)
{
    GameObject cellObject = HFEngine::ECS.CreateGameObject(parent, std::to_string(cell->index()));
    HFEngine::ECS.GetComponent<Transform>(cellObject).SetPosition({cell->coord()->x, 0.0f, cell->coord()->y});

    MapCell mCell;
    mCell.CellSiteIndex = cell->index();
    mCell._BaseDelaunayPolygon = CreateLocalPolygon(cell, bounds);
    HFEngine::ECS.AddComponent<MapCell>(cellObject, mCell);
    return cellObject;
}

void MapGenerator::CreateBridges(std::vector<GameObject> cells, GameObject parent)
{
    auto GetCell = [cells](Delaunay::Site* s) {
        for (GameObject cell : cells)
            if (HFEngine::ECS.GetComponent<MapCell>(cell).CellSiteIndex == s->index())
                return cell;
        return (GameObject)0;
    };

    for (auto edge : edges)
    {
        if (!edge->visible())
            continue;

        Delaunay::Point* leftVertex = edge->clippedEnds()[Delaunay::LR::LEFT];
        Delaunay::Point* rightVertex = edge->clippedEnds()[Delaunay::LR::RIGHT];

        if (Delaunay::Point::distance(leftVertex, rightVertex) < config.minEdgeLengthForBridge)
            continue;

        GameObject cellA = GetCell(edge->leftSite());
        GameObject cellB = GetCell(edge->rightSite());
        if (cellA == (GameObject)0 || cellB == (GameObject)0)
            continue;

        glm::vec3 position = glm::vec3(
            (leftVertex->x + rightVertex->x) * 0.5f,
            0.0f,
            (leftVertex->y + rightVertex->y) * 0.5f
            );

        float rotation = rad2deg(glm::atan(rightVertex->x - position.x, rightVertex->y - position.z)) + 90; // + 90 for right
        GameObject bridgeObject = config.bridgePrefab->Instantiate(parent, position, { 0.0f, rotation, 0.0f });

        CellBridge bridge;
        bridge.CellA = cellA;
        bridge.CellB = cellB;
        HFEngine::ECS.AddComponent<CellBridge>(bridgeObject, bridge);

        HFEngine::ECS.GetComponent<MapCell>(cellA).Bridges.push_back({ bridgeObject, cellB, (GameObject)0 });
        HFEngine::ECS.GetComponent<MapCell>(cellB).Bridges.push_back({ bridgeObject, cellA, (GameObject)0 });
    }
}


bool MapGenerator::ValidateMapLayout(std::vector<GameObject> cells)
{
    int cellsWithOneBridgeCount = 0;
    for (auto cell : cells)
    {
        MapCell mc = HFEngine::ECS.GetComponent<MapCell>(cell);
        if (mc.Bridges.size() == 0)
        {
            return false; // invalid cell with no bridges, throw away whole map...
        }
        else if (mc.Bridges.size() == 1)
        {
            cellsWithOneBridgeCount++;
        }
    }

    // there must be at least one cell with only one bridge
    if (cellsWithOneBridgeCount >= 1)
        return true;
    else
        return false;
}
